### PR TITLE
Add duplicate map leaves test to sequencer.

### PR DESCRIPTION
**Pull request by gdbelvin**
Adds a test to the KT sequencer to ensure that it never sends a `tlog.SetLeaves` request with duplicate `MapLeaf.Index`s. 

Note that this did not detect any problems with sending duplicate map leaves to the Trillian Map, so the root case of the error `SetLeaves(): CalculateRoot(): Error 1062: Duplicate entry '3584314497260755511-@-25483' for key 'PRIMARY'"` is still unknown

### CODE REVIEW

File: core/sequencer/sequencer.go [hunk 1]
```diff
@@ -239,12 +239,14 @@ func (s *Sequencer) applyMutations(mutations []*mutator.QueueMessage, leaves []*
...
 		// Serialize commitment.
-		extraData, err := proto.Marshal(m.ExtraData)
+		extraData, err := proto.Marshal(msg.ExtraData)
 		if err != nil {
 			glog.Warningf("Marshal(committed proto): %v", err)
 			continue
 		}
 
+		// Make sure that only ONE MapLeaf is output per index.
+		// Failure to do so will corrupt the map.
```
Comment (daviddrysdale) [inline core/sequencer/sequencer.go]: "corrupt the map" is (I think) too strong -- maybe "... will result in a rejected request".

File: core/sequencer/sequencer_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+	return &mutator.QueueMessage{
+		ID:        id,
+		Mutation:  update.EntryUpdate.Mutation,
+		ExtraData: update.EntryUpdate.Committed,
+	}
+}
+
+// TestDuplicateMutations verifies that each call to tlog.SetLeaves specifies
+// each mapleaf.Index at most ONCE.  Failure to do so will corrupt the map.
```
Comment (daviddrysdale) [inline core/sequencer/sequencer_test.go]: comment too strong (as above)?

File: core/sequencer/sequencer_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+func queueMsg(id int64, signer *tink.KeysetHandle) *mutator.QueueMessage {
+	index := []byte{byte(id)}
+	userID := string(id)
+	m := entry.NewMutation(index, "domain", "app", userID)
+	signers := []*tink.KeysetHandle{signer}
+	pubkey, err := signer.GetPublicKeysetHandle()
+	if err != nil {
+		panic(fmt.Errorf("GetPublicKeysetHandle(): %v", err))
```
Comment (daviddrysdale) [inline core/sequencer/sequencer_test.go]: Alternatively, you could pass in `t`, make this a `t.Helper()` and do `t.Fatal` here -- up to you.

File: core/sequencer/sequencer_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		msgs       []*mutator.QueueMessage
+		leaves     []*tpb.MapLeaf
+		wantLeaves int
+	}{
+		{
+			desc: "duplicate",
+			msgs: []*mutator.QueueMessage{
+				queueMsg(1, keyset),
+				queueMsg(1, keyset),
+			},
```
Comment (daviddrysdale) [inline core/sequencer/sequencer_test.go]: BTW, what happens if you had a testcase with `queueMsg(1, keyset), queueMsg(1,keyset2)`?

File: core/sequencer/sequencer.go (hunk 1 above)
Comment (gdbelvin) [reply core/sequencer/sequencer.go]: Can you confirm that the mapserver will reject the request without processing it? 
Our previous theory was that duplicate leaves were creating errors deep in the map server.

File: core/sequencer/sequencer_test.go (hunk 4 above)
Comment (gdbelvin) [reply core/sequencer/sequencer_test.go]: I'll add this test case. 

What *should* happen, is the sequencer will randomly pick only one of them to succeed.

File: core/sequencer/sequencer.go (hunk 1 above)
Comment (daviddrysdale) [reply core/sequencer/sequencer.go]: [It should do now](https://github.com/google/trillian/commit/be4e023c1707d2c6e836d97818d37501d910d5bb)

Comment (gdbelvin) [reply core/sequencer/sequencer.go]: Thanks

File: core/sequencer/sequencer_test.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	t.Helper()
+	index := []byte{byte(id)}
+	userID := string(id)
+	m := entry.NewMutation(index, "domain", "app", userID)
+	signers := []*tink.KeysetHandle{signer}
+	pubkey, err := signer.GetPublicKeysetHandle()
+	if err != nil {
+		t.Fatalf("GetPublicKeysetHandle(): %v", err)
+	}
+	m.ReplaceAuthorizedKeys(pubkey.Keyset())
```
Comment (golangci) [inline core/sequencer/sequencer_test.go]: Error return value of `m.ReplaceAuthorizedKeys` is not checked

File: core/sequencer/sequencer_test.go (hunk 2 above)
Comment (gdbelvin) [reply core/sequencer/sequencer_test.go]: done

File: core/sequencer/sequencer_test.go (hunk 3 above)
Comment (gdbelvin) [reply core/sequencer/sequencer_test.go]: done

File: core/sequencer/sequencer_test.go (hunk 2 above)
Comment (gdbelvin) [reply core/sequencer/sequencer_test.go]: fixed.

File: core/mutator/entry/mutation.go [hunk 6]
```diff
@@ -4,8 +4,7 @@
 // you may not use this file except in compliance with the License.
 // You may obtain a copy of the License at
 //
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
+//     http://www.apache.org/licenses/LICENSE-2.0 //
```
Comment (daviddrysdale) [inline core/mutator/entry/mutation.go]: Errant line join

File: core/sequencer/sequencer_test.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	}
+
+	for _, tc := range []struct {
+		desc       string
+		msgs       []*mutator.QueueMessage
+		leaves     []*tpb.MapLeaf
+		wantLeaves int
+	}{
+		{
+			desc: "duplicate",
```
Comment (daviddrysdale) [inline core/sequencer/sequencer_test.go]: Maybe have two testcases, one with same values and one with different (ie. have both the old case and the current cast) ?

File: core/mutator/entry/mutation.go (hunk 6 above)
Comment (gdbelvin) [reply core/mutator/entry/mutation.go]: thx

File: core/sequencer/sequencer_test.go (hunk 7 above)
Comment (gdbelvin) [reply core/sequencer/sequencer_test.go]: cool

**Pull request merged**